internal/upload/azure/azurestorage.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	// Hash the imageFile
+	imageFileHash := md5.New()
+	if _, err := io.Copy(imageFileHash, imageFile); err != nil {
+		return fmt.Errorf("cannot create md5 of the image: %v", err)
+	}
+	// Move the cursor back to the start of the imageFile
+	if _, err := imageFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("cannot seek the image: %v", err)
+	}
+
+	// Create page blob URL. Page blob is required for VM images
+	blobURL := newPageBlobURL(containerURL, metadata.BlobName)
+	_, err = blobURL.Create(ctx, stat.Size(), 0, azblob.BlobHTTPHeaders{}, azblob.Metadata{}, azblob.BlobAccessConditions{})
+	if err != nil {
+		return fmt.Errorf("cannot create the blob URL: %v", err)
+	}
+	// Wrong MD5 does not seem to have any impact on the upload
+	_, err = blobURL.SetHTTPHeaders(ctx, azblob.BlobHTTPHeaders{ContentMD5: imageFileHash.Sum(nil)}, azblob.BlobAccessConditions{})
+	if err != nil {
+		return fmt.Errorf("cannot set the HTTP headers on the blob URL: %v", err)
+	}
+
+	// Create control variables
+	// This channel simulates behavior of a semaphore and bounds the number of parallel threads
+	var semaphore = make(chan int, threads)
+	// Forward error from goroutine to the caller
+	var errorInGoroutine = make(chan error, 1)
+	var counter int64 = 0
+
+	// Create buffered reader to speed up the upload
+	reader := bufio.NewReader(imageFile)
+	// Run the upload
+	run := true
+	var wg sync.WaitGroup
+	for run {
+		buffer := make([]byte, azblob.PageBlobMaxUploadPagesBytes)
+		n, err := reader.Read(buffer)
+		if err != nil {
+			if err == io.EOF {
+				run = false
+			} else {
+				return fmt.Errorf("reading the image failed: %v", err)
+			}
+		}
+		if n == 0 {
+			break
+		}
+		wg.Add(1)
+		semaphore <- 1
+		go func(counter int64, buffer []byte, n int) {
+			defer wg.Done()
+			_, err = blobURL.UploadPages(ctx, counter*azblob.PageBlobMaxUploadPagesBytes, bytes.NewReader(buffer[:n]), azblob.PageBlobAccessConditions{}, nil)
+			if err != nil {
+				err = fmt.Errorf("uploading a page failed: %v", err)
+				// Send the error to the error channel in a non-blocking way. If there is already an error, just discard this one
+				select {
+				case errorInGoroutine <- err:
+				default:
+				}
+			}
+			<-semaphore
+		}(counter, buffer, n)
+		counter++
+	}
+	// Wait for all goroutines to finish
+	wg.Wait()
+	// Check any errors during the transmission using a nonblocking read from the channel
+	select {
+	case err := <-errorInGoroutine:
+		return err
+	default:
+	}
+	// Check properties, specifically MD5 sum of the blob
+	props, err := blobURL.GetProperties(ctx, azblob.BlobAccessConditions{})
+	if err != nil {
+		return fmt.Errorf("getting the properties of the new blob failed: %v", err)
+	}
+	var blobChecksum []byte = props.ContentMD5()
+	var fileChecksum []byte = imageFileHash.Sum(nil)
+
+	if !bytes.Equal(blobChecksum, fileChecksum) {
+		return errors.New("error during image upload. the image seems to be corrupted")
```

Comment:
I implemented a similar check for GCP, but I'm thinking that maybe the uploaded file should be deleted from the storage if the checksum check fails. Not an issue for this set of changes, but I would like to get your point of view anyway.

Reply:
Good point, we should open an issue to track it.